### PR TITLE
Get Issue#8 #9 tag validation and default db migration done 

### LINKED ISSUE #8: Required Tag Association
#### As a user I want all images to have at least one tag.

__Story__:

The tag feature has been a hit. Our users love being able to filter images by
cats, SR-71s, and "AppFolio tshirts" to name a few. In fact, we have discovered
that images without any tags receive significantly less views than those with
tags. As a result, we have decided to require all images to have at least one
tag.

__Acceptance criteria__:
- [x] I cannot save an image without adding at least one tag.
- [x] An error message is associated with the appropriate input field when a
tag is not provided.
- [x] All images have at least one tag.

__Discussion Topic__:
- [x] What do you do about existing images that have no tags? 
	- [x] Pretend this is a production system that has been running for years with thousands of users 
	and zillions of images. You can't just drop your production db and rebuild it. Your solution should consider the following points:
		- [x] You can't just delete the tagless images.
		- [x] You shouldn't annoy users	by presenting them with validation errors if they try to open or edit a tagless image.
		- [x] If your solution modifies your production db, then your solution should be able to un-modify your production db 
			  if you decide to run an older version of your code on the production db. (Hint: what Rails mechanism performs
			  versioned db changes?)

__Dependencies__:
- Image Tags
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by limbov8**
Closes #8 #9 

#9 Modify Tag List, already done before.

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: db/migrate/20190523011727_add_default_tag_for_existing_images.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class AddDefaultTagForExistingImages < ActiveRecord::Migration[5.2]
+  def up
+    tag = ActsAsTaggableOn::Tag.find_by(name: 'Default')
+    tag = ActsAsTaggableOn::Tag.create(name: 'Default') if tag.nil? || tag.invalid?
+    #  => #<ActsAsTaggableOn::Tagging id: 2, tag_id: 2, taggable_type: "Image", taggable_id: 21,
+    # tagger_type: nil, tagger_id: nil, context: "tags", created_at: "2019-05-22 17:50:05">
+    Image.all.each do |img|
```
Comment (choruk) [inline db/migrate/20190523011727_add_default_tag_for_existing_images.rb]: It is generally considered a bad practice to reference class names inside of a migration. Given the current state of your app's code, this migration works, but what happens if you want to rename the image class? Or if you rename the `tag_list` method? Or in the case of the `ActsAsTaggableOn::Tagging` class, what happens if the code for the class changes in such a way that it no longer knows about the `context` column? Basically, your models will reflect the current state of the schema / codebase, but your migration will be running with the version of the schema that existed at that point in time (if that makes sense). As such, it is better to stick to writing SQL directly in these types of migrations. It makes things a bit more complicated, but is more future proof / resilient to changes in your application code.

If you are thinking "well I can just edit the migration if it becomes an issue" that is also problematic and is generally considered a bad practice to edit migrations that have already been released to production. Once the migration runs in production, it will never run again, so any changes you make to it will only be reflected for newer customers/sites that are getting created. If you edit the migration, you could introduce a slight change in behavior between the two versions and end up with slightly different looking DB (not great to have inconsistencies at the DB level).

So for example instead of:
```ruby
tag = ActsAsTaggableOn::Tag.find_by(name: 'Default')
tag = ActsAsTaggableOn::Tag.create(name: 'Default') if tag.nil? || tag.invalid?
```
You would do:
```ruby
tag_id = execute <<-SQL
  SELECT id FROM tags WHERE name = 'Default'
SQL
if tag_id.first.nil?
  execute <<-SQL
    INSERT INTO tags (name, created_at, updated_at) VALUES ('Default', UTC_TIMESTAMP(), UTC_TIMESTAMP())
  SQL
  tag_id = execute <<-SQL
    SELECT id FROM tags WHERE name = 'Default'
  SQL
end
tag_id = tag_id.first.first
```

File: db/migrate/20190521161831_create_images.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 class CreateImages < ActiveRecord::Migration[5.2]
   def change
     create_table :images do |t|
-      t.text :link, null: false
+      t.text :link
```
Comment (choruk) [inline db/migrate/20190521161831_create_images.rb]: Was this change intentional?

File: db/migrate/20190523011727_add_default_tag_for_existing_images.rb [hunk 3]
```diff
@@ -1,20 +1,48 @@
...
+        INSERT INTO tags (name) VALUES ('Default')
+      SQL
+      tag_id = execute <<-SQL
+        SELECT id FROM tags WHERE name = 'Default'
+      SQL
+    end
+    tag_id = tag_id.first['id']
+
+    images = execute <<-SQL
+      SELECT images.id from images where images.id not in (SELECT taggings.taggable_id from taggings where taggings.tag_id = #{tag_id} and taggings.taggable_type = 'Image' and taggings.context = 'tags')
```
Comment (choruk) [inline db/migrate/20190523011727_add_default_tag_for_existing_images.rb]: This query will actually return the IDs of all images that aren't tagged with the `Default` tag, but I think you actually want the IDs of all images that aren't tagged with *any* tag, right?

File: db/migrate/20190521161831_create_images.rb (hunk 2 above)
Comment (limbov8) [reply db/migrate/20190521161831_create_images.rb]: Yes, this one is intentional since all the tests were not  considering this database validation. Otherwise, all the tests need to be rewritten. : (

File: db/seeds.rb [hunk 4]
```diff
@@ -6,23 +6,28 @@
...
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2016/08/26/1012523/2d5e8e09c3f7c4d0695ae7a8a11d4008-500.jpg' },
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2013/03/21/55890/935d4c0de28003f04f75185eefa96b50-500.jpg' },
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2014/02/25/276134/b42683d82bca8998e28868a3f0c774a8-500.jpg' },
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2013/06/11/100995/1fd0b206217b7cddd841a43936bfb0c3-500.jpg' },
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2013/04/14/75364/db9e8ccd0267251066eaab0adb1f7b86-500.jpg' },
+ { link: 'https://www.wallpaperup.com/uploads/wallpapers/2013/09/12/147040/a9c08e9464538e237f709d812b51ce99-500.jpg' }]
+
+image_seeds.each do |img|
+  images = Image.new(img)
+  images.save!(validate: false)
```
Comment (choruk) [inline db/seeds.rb]: You should be careful about using `validate: false` when saving, as it bypasses all validations. As a result, you are actually generating 20 invalid images from this seeds file, which seeds your DB with invalid data. The real fix is to make sure the images are valid when being created.
